### PR TITLE
chore(ci): block new AIDEV comment anchors

### PR DESCRIPTION
**Pull request by vlad-scherbich**
## Description

Guild decision: stop using `AIDEV-NOTE:` / `AIDEV-TODO:` / `AIDEV-QUESTION:` prefixes in favor of plain inline comments.

This PR only adds the policy and the gate:

- Update `AGENTS.md` and cursor rules so new anchors are not added
- Add `scripts/check_no_new_aidev_anchors.py` (diff-scoped: added lines only; legacy anchors are grandfathered)
- Wire a GitLab CI job modeled on `check_added_file_size` (includes `GIT_DEPTH: "0"` to match that job on current `main`)
- Exclude the new script from `pyproject.toml` `exclude_dirs`

## Testing

- [x] Pre-commit hooks passed on this branch
- [ ] Confirm `check_no_new_aidev_anchors` runs on the MR and is skippable on `main` / release / merge-queue
- [ ] `python3 scripts/check_no_new_aidev_anchors.py --base-ref origin/main` is clean on this branch (no new anchors in this diff)

## Additional Notes

https://datadoghq.atlassian.net/browse/PROF-16007

### CODE REVIEW

File: scripts/check_no_new_aidev_anchors.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+from __future__ import annotations
+
+import argparse
+import re
+import subprocess  # nosec B404
+import sys
+from typing import Optional
+
+
+# Match Python, C/C++, and Rust comment introducers only — avoids false
```
Comment (brettlangdon) [inline scripts/check_no_new_aidev_anchors.py]: why do we need to search specifically for a comment?

or even specific suffix, seems a grep on `AIDEV` on the diff should be good enough? as long as we are searching only on additions/changes and not deletions?


tl;dr; I wouldn't worry about making it too complicated, would rather have false positives than negatives.

**Review by brettlangdon — APPROVED** [review]
this could make sense as a targeted GHA. since it isn't really about poor code quality, just consistency, and if it doesn't run on non-PRs, then it shouldn't really matter much (vs, what if it is flaky/fails on a non-pr run?)

feedback/suggestions, but nothing really wrong with current implementation.

File: scripts/check_no_new_aidev_anchors.py (hunk 1 above)
Comment (vlad-scherbich) [reply scripts/check_no_new_aidev_anchors.py]: Makes sense, will change.

**Comment by vlad-scherbich** [comment]
> this could make sense as a targeted GHA. since it isn't really about poor code quality, just consistency, and if it doesn't run on non-PRs, then it shouldn't really matter much (vs, what if it is flaky/fails on a non-pr run?)
> 
> feedback/suggestions, but nothing really wrong with current implementation.

@brettlangdon agreed - will add this separately: https://github.com/DataDog/dd-trace-py/pull/20211

**Review by taegyunkim — APPROVED** [review]

File: scripts/check_no_new_aidev_anchors.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import argparse
+import re
+import subprocess  # nosec B404
+import sys
+from typing import Optional
+
+
+# Deliberately match any occurrence in added lines. False positives are
+# preferable to allowing a new deprecated label to bypass the check.
+ANCHOR_RE: re.Pattern[str] = re.compile(r"AIDE" r"V")
```
[the remark this answers is not in the document]
Comment (vlad-scherbich) [reply scripts/check_no_new_aidev_anchors.py:24]: Addressed in c283b15e

File: AGENTS.md [hunk 3]
```diff
@@ -139,3 +130,15 @@ Use the Skill tool to invoke these. **Always prefer skills over raw commands.**
 | Repository Structure                                            | `.cursor/rules/repo-structure.mdc`                                      | —                                                                                                                                                   |
 | Linting                                                         | `.cursor/rules/linting.mdc`                                             | —                                                                                                                                                   |
 | Testing                                                         | `.cursor/rules/testing.mdc`                                             | —                                                                                                                                                   |
+
+## AIDE&#x56; Anchor Comments
```
Comment (KowalskiThomas) [inline AGENTS.md:134]: That's smart 🤣

**Review by KowalskiThomas — APPROVED** [review]

**Review by gnufede — APPROVED** [review]

**Comment by vlad-scherbich** [comment]
> ### 🟡 Changes recommended
> Unresolved credential exposure risk and checker bypasses block approval.
> 
> _Get a fresh assessment by requesting another Copilot review._
> 
> Pull request overview
> Adds policy and CI enforcement to prevent new `AIDEV-*` comment anchors while preserving legacy anchors.
> 
> **Changes:**
> 
> * Adds diff-scoped checker and tests.
> * Adds GitLab and GitHub CI gates.
> * Updates repository and Cursor guidance.
> 
> File summaries
> File	Summary
> `tests/suitespec.yml`	Maps checker coverage to the internal suite.
> `tests/internal/test_check_no_new_aidev_anchors.py`	Adds checker tests; positive added-line and `+++` edge-case coverage is needed.
> `scripts/check_no_new_aidev_anchors.py`	Has critical parser bypasses involving `+++` lines, apostrophes, and inline docstrings; description also promises an absent `pyproject.toml` exclusion.
> `AGENTS.md`	Documents the deprecated-anchor policy.
> `.gitlab-ci.yml`	Critical: PR-controlled checker code runs with GitHub/OIDC credentials available.
> `.github/workflows/check-no-new-aidev-anchors.yml`	Adds the GitHub pull request check.
> `.cursor/rules/dd-trace-py.mdc`	Updates general agent guidance.
> `.cursor/rules/ai-guard.mdc`	Updates AI Guard guidance.
> Review details
> ### Suppressed comments (3)
> **scripts/check_no_new_aidev_anchors.py:54**
> 
> * An added line whose content starts with `++` is emitted by Git as `+++...`; this condition skips it before `_is_anchor_line` runs. Since the checker intentionally scans any unquoted `AIDEV` occurrence, such a line can bypass the gate. Only the `+++ b/...` header handled above should be excluded.
> 
> ```
>         if not line.startswith("+") or line.startswith("+++"):
>             continue
> ```
> 
> **scripts/check_no_new_aidev_anchors.py:25**
> 
> * The quote-stripping regex also runs over comment text, so an apostrophe before an anchor can be paired with a later apostrophe and remove the real violation. For example, `# It's an AIDEV-NOTE: user's token` is stripped through the second apostrophe and passes this check. Use a comment-aware scanner or otherwise avoid treating arbitrary apostrophes in comments as string delimiters.
> 
> ```
> STRING_RE: re.Pattern[str] = re.compile(r"""(["'`])(?:\\.|(?!\1).)*\1""")
> ```
> 
> **scripts/check_no_new_aidev_anchors.py:65**
> 
> * One-line Python docstrings bypass this check. For example, an added `def f(): "AIDEV-NOTE: ..."` is a real docstring anchor, but `TRIPLE_STRING_RE` only handles lines whose first token is a triple quote and `STRING_RE.sub` then removes the quoted text. Since the tests treat docstrings as anchors, parse Python docstrings or explicitly handle inline function/class docstrings so this cannot bypass the gate.
> 
> ```
>     if TRIPLE_STRING_RE.search(content) is not None:
>         return True
>     content_without_strings: str = STRING_RE.sub("", content)
>     return ANCHOR_RE.search(content_without_strings) is not None
> ```
> 
> * **Files reviewed:** 8/8 changed files
> * **Comments generated:** 3
> * **Review effort level:** Lite
> 
> 💡 [Add a `code-review` agent skill](/DataDog/dd-trace-py/new/main?filename=.github/skills/code-review/SKILL.md) or configure MCP servers for context-aware, tailored reviews. [Learn more in the docs.](https://docs.github.com/copilot/how-tos/use-copilot-agents/request-a-code-review/use-code-review?tool=webui#mcp-servers-and-agent-skills)

@copilot-pull-request-reviewer addressed concerns

image

**Review by avara1986 — APPROVED** [review]
